File: src/lib/studio/transcription-dictionary.ts
import type { RawWord } from "@/lib/studio/transcribe-remote";

export const MAX_DICTIONARY_ENTRIES = 100;
export const MAX_DICTIONARY_TERM_LENGTH = 80;
export const MAX_DICTIONARY_ALIASES = 20;

export interface TranscriptionDictionaryEntry {
  id: string;
  term: string;
  aliases: string[];
}

export interface CaptionCorrection {
  heard: string;
  term: string;
}

/** Case/punctuation-insensitive key used for matching and uniqueness. */
export function dictionaryKey(value: string): string {
  return value
    .normalize("NFKC")
    .toLocaleLowerCase("en")
    .replace(/[^\p{L}\p{N}]+/gu, "")
    .slice(0, MAX_DICTIONARY_TERM_LENGTH);
}

export function cleanDictionaryValue(value: string): string {
  return value.trim().replace(/\s+/g, " ").slice(0, MAX_DICTIONARY_TERM_LENGTH);
}

export function cleanDictionaryAliases(values: string[]): string[] {
  const seen = new Set<string>();
  const aliases: string[] = [];
  for (const value of values) {
    const alias = cleanDictionaryValue(value);
    const key = dictionaryKey(alias);
    if (!key || seen.has(key)) continue;
    seen.add(key);
    aliases.push(alias);
    if (aliases.length >= MAX_DICTIONARY_ALIASES) break;
  }
  return aliases;
}

/**
 * A conservative caption diff for the inline "remember this" affordance.
 * It only fires for a one-token substitution, never for rewrites, additions,
 * or punctuation-only edits where the user's intent would be ambiguous.
 */
export function findCaptionCorrection(
  before: string,
  after: string,
): CaptionCorrection | null {
  const from = before.trim().split(/\s+/).filter(Boolean);
  const to = after.trim().split(/\s+/).filter(Boolean);
  if (from.length === 0 || from.length !== to.length) return null;

  let correction: CaptionCorrection | null = null;
  for (let i = 0; i < from.length; i++) {
    if (dictionaryKey(from[i]) === dictionaryKey(to[i])) continue;
    if (correction) return null;
    const heard = cleanDictionaryValue(
      from[i].replace(/^[^\p{L}\p{N}]+|[^\p{L}\p{N}]+$/gu, ""),
    );
    const term = cleanDictionaryValue(
      to[i].replace(/^[^\p{L}\p{N}]+|[^\p{L}\p{N}]+$/gu, ""),
    );
    if (!dictionaryKey(heard) || !dictionaryKey(term)) return null;
    correction = { heard, term };
  }
  return correction;
}

const trailingPunctuation = (value: string) =>
  value.match(/[,.!?;:]+$/u)?.[0] ?? "";

interface Pattern {
  entry: TranscriptionDictionaryEntry;
  tokens: string[];
}

/**
 * Apply only user-approved spellings. Canonical terms are included as patterns
 * so a lowercase ASR result still receives the user's exact casing. Multi-word
 * aliases retain per-word timings when token counts match; otherwise they
 * collapse to the full acoustic range instead of inventing timestamps.
 */
export function applyTranscriptionDictionary(
  words: RawWord[],
  entries: TranscriptionDictionaryEntry[],
): RawWord[] {
  if (words.length === 0 || entries.length === 0) return words;

  const patterns: Pattern[] = [];
  for (const entry of entries) {
    for (const value of [entry.term, ...entry.aliases]) {
      const tokens = value.split(/\s+/).map(dictionaryKey).filter(Boolean);
      if (tokens.length > 0) patterns.push({ entry, tokens });
    }
  }
  patterns.sort((a, b) => b.tokens.length - a.tokens.length);

  const out: RawWord[] = [];
  for (let i = 0; i < words.length; ) {
    const match = patterns.find(({ tokens }) =>
      tokens.every(
        (token, offset) =>
          dictionaryKey(words[i + offset]?.text ?? "") === token,
      ),
    );
    if (!match) {
      out.push(words[i]);
      i += 1;
      continue;
    }

    const source = words.slice(i, i + match.tokens.length);
    const canonical = match.entry.term.split(/\s+/).filter(Boolean);
    if (canonical.length === source.length) {
      for (let j = 0; j < source.length; j++) {
        out.push({
          ...source[j],
          text:
            canonical[j] +
            (j === source.length - 1
              ? trailingPunctuation(source[j].text)
              : ""),
        });
      }
    } else {
      out.push({
        text:
          match.entry.term +
          trailingPunctuation(source[source.length - 1].text),
        start: source[0].start,
        end: source[source.length - 1].end,
      });
    }
    i += match.tokens.length;
  }
  return out;
}

/** Terms sent to the ASR, deduped and bounded for a safe request URL. */
export function dictionaryKeyterms(
  entries: TranscriptionDictionaryEntry[],
): string[] {
  const seen = new Set<string>();
  const terms: string[] = [];
  let requestChars = 0;
  for (const entry of entries) {
    const term = cleanDictionaryValue(entry.term);
    const key = dictionaryKey(term);
    if (!key || seen.has(key)) continue;
    // Keep the query safely below common proxy URL limits. Every saved entry
    // still participates in deterministic alias correction after ASR.
    if (requestChars + term.length > 3_000) break;
    seen.add(key);
    terms.push(term);
    requestChars += term.length;
    if (terms.length >= 50) break;
  }
  return terms;
}

File: native-macos/Sources/YapperNative/Models/Dictionary/TranscriptionDictionary.swift
import Foundation

/// A spelling the creator has approved, and the ways transcription tends to get
/// it wrong.
struct DictionaryEntry: Codable, Equatable, Identifiable, Sendable {
    let id: String
    var term: String
    var aliases: [String]

    init(id: String = "local-\(UUID().uuidString)", term: String, aliases: [String] = []) {
        self.id = id
        self.term = TranscriptionDictionary.cleanValue(term)
        self.aliases = TranscriptionDictionary.cleanAliases(aliases)
    }
}

/// A one-word fix the creator made to a caption, which is worth remembering.
struct CaptionCorrection: Equatable, Sendable {
    let heard: String
    let term: String
}

/// The creator's own spellings, applied to what the transcriber heard.
///
/// Only approved spellings are ever applied: the canonical term is matched too,
/// so a lowercase result still picks up the creator's own capitalisation.
enum TranscriptionDictionary {
    static let maximumEntries = 100
    static let maximumTermLength = 80
    static let maximumAliases = 20

    /// The form two spellings are compared in: letters and numbers only, so
    /// case and punctuation cannot make two of the same word.
    static func key(_ value: String) -> String {
        String(
            value
                .precomposedStringWithCompatibilityMapping
                .lowercased()
                .filter { $0.isLetter || $0.isNumber }
                .prefix(maximumTermLength)
        )
    }

    static func cleanValue(_ value: String) -> String {
        String(
            value
                .trimmingCharacters(in: .whitespacesAndNewlines)
                .split(whereSeparator: \.isWhitespace)
                .joined(separator: " ")
                .prefix(maximumTermLength)
        )
    }

    static func cleanAliases(_ values: [String]) -> [String] {
        var seen: Set<String> = []
        var aliases: [String] = []
        for value in values {
            let alias = cleanValue(value)
            let key = key(alias)
            guard !key.isEmpty, !seen.contains(key) else { continue }
            seen.insert(key)
            aliases.append(alias)
            if aliases.count >= maximumAliases { break }
        }
        return aliases
    }

    /// The terms worth telling the transcriber about up front, deduped and
    /// bounded so the request URL stays a sane length. Everything saved still
    /// takes part in the correction pass afterwards.
    static func keyterms(_ entries: [DictionaryEntry]) -> [String] {
        var seen: Set<String> = []
        var terms: [String] = []
        var characters = 0
        for entry in entries {
            let term = cleanValue(entry.term)
            let key = key(term)
            guard !key.isEmpty, !seen.contains(key) else { continue }
            if characters + term.count > 3_000 { break }
            seen.insert(key)
            terms.append(term)
            characters += term.count
        }
        return terms
    }

    /// The words as the creator spells them.
    ///
    /// A multi-word alias keeps per-word timings when the word counts match;
    /// otherwise it collapses onto the whole stretch it replaced rather than
    /// inventing timestamps for words nobody said separately.
    static func applied(
        to words: [TranscriptWord],
        entries: [DictionaryEntry]
    ) -> [TranscriptWord] {
        guard !words.isEmpty, !entries.isEmpty else { return words }

        struct Pattern {
            let entry: DictionaryEntry
            let tokens: [String]
        }
        var patterns: [Pattern] = []
        for entry in entries {
            for value in [entry.term] + entry.aliases {
                let tokens = value
                    .split(whereSeparator: \.isWhitespace)
                    .map { key(String($0)) }
                    .filter { !$0.isEmpty }
                if !tokens.isEmpty { patterns.append(Pattern(entry: entry, tokens: tokens)) }
            }
        }
        // Longest first, so "New York City" wins over "New York".
        patterns.sort { $0.tokens.count > $1.tokens.count }

        var result: [TranscriptWord] = []
        var index = 0
        while index < words.count {
            let match = patterns.first { pattern in
                guard index + pattern.tokens.count <= words.count else { return false }
                return pattern.tokens.enumerated().allSatisfy { offset, token in
                    key(words[index + offset].text) == token
                }
            }
            guard let match else {
                result.append(words[index])
                index += 1
                continue
            }

            let source = Array(words[index ..< index + match.tokens.count])
            let canonical = match.entry.term
                .split(whereSeparator: \.isWhitespace)
                .map(String.init)
            if canonical.count == source.count {
                for (offset, word) in source.enumerated() {
                    var corrected = word
                    corrected.text = canonical[offset]
                        + (offset == source.count - 1 ? trailingPunctuation(word.text) : "")
                    result.append(corrected)
                }
            } else if let first = source.first, let last = source.last {
                result.append(
                    TranscriptWord(
                        mediaID: first.mediaID,
                        text: match.entry.term + trailingPunctuation(last.text),
                        start: first.start,
                        end: last.end
                    )
                )
            }
            index += match.tokens.count
        }
        return result
    }

    /// A caption edit worth remembering: exactly one word swapped for another.
    /// Rewrites, insertions and punctuation-only changes are left alone,
    /// because what the creator meant by them is anyone's guess.
    ///
    /// A change of capitals counts. Fixing `celpip` to `CELPIP` is the whole
    /// point of a dictionary entry — the saved term is what supplies the
    /// capitals to every transcript after it.
    static func correction(before: String, after: String) -> CaptionCorrection? {
        let from = before.split(whereSeparator: \.isWhitespace).map(String.init)
        let to = after.split(whereSeparator: \.isWhitespace).map(String.init)
        guard !from.isEmpty, from.count == to.count else { return nil }

        var correction: CaptionCorrection?
        for (index, word) in from.enumerated() {
            let heard = cleanValue(stripped(word))
            let term = cleanValue(stripped(to[index]))
            // Punctuation is not a spelling, so a comma appearing is no change.
            guard heard != term else { continue }
            guard correction == nil else { return nil }
            guard !key(heard).isEmpty, !key(term).isEmpty else { return nil }
            correction = CaptionCorrection(heard: heard, term: term)
        }
        return correction
    }

    private static func trailingPunctuation(_ value: String) -> String {
        String(value.reversed().prefix { ",.!?;:".contains($0) }.reversed())
    }

    private static func stripped(_ value: String) -> String {
        var trimmed = Substring(value)
        while let first = trimmed.first, !first.isLetter, !first.isNumber {
            trimmed = trimmed.dropFirst()
        }
        while let last = trimmed.last, !last.isLetter, !last.isNumber {
            trimmed = trimmed.dropLast()
        }
        return String(trimmed)
    }
}
